qiskit/scheduler/methods/sequence.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Mapping a scheduled QuantumCircuit to a pulse Schedule.
+"""
+import warnings
+from collections import defaultdict
+
+from qiskit.circuit.barrier import Barrier
+from qiskit.circuit.measure import Measure
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.pulse.schedule import Schedule
+from qiskit.pulse.transforms import pad
+
+from qiskit.scheduler.config import ScheduleConfig
+from qiskit.scheduler.methods.lowering import lower_gates
+
+
+def sequence(scheduled_circuit: QuantumCircuit, schedule_config: ScheduleConfig) -> Schedule:
+    """
+    Return the pulse Schedule which implements the input scheduled circuit.
+
+    Assume all measurements are done at once at the last of the circuit.
+    Schedules according to the command definition given by the schedule_config.
+
+    Args:
+        scheduled_circuit: The scheduled quantum circuit to translate.
+        schedule_config: Backend specific parameters used for building the Schedule.
+
+    Returns:
+        A schedule corresponding to the input ``circuit``.
+    """
+    # recall start times
+    qubit_time_available = defaultdict(int)
+    start_times = []
+    measure_times = []
+    for inst, qubits, _ in scheduled_circuit.data:
+        start_time = qubit_time_available[qubits[0]]
+        if isinstance(inst, Measure):
+            measure_times.append(start_time)
+        else:
+            start_times.append(start_time)
+        duration = scheduled_circuit.instruction_durations.get(inst, qubits)
```

Comment:
So, regarding the conversation we started on `schedule.py`, why does this have to be from `instruction_durations`? Is it to do error checking against the user defined durations?

The alternative way to do this would be like the other methods. Below, you do:
```
    circ_pulse_defs = lower_gates(scheduled_circuit, schedule_config)
```
Each CircuitPulseDef in this list has an explicit duration. I would make `instruction_durations` optional and check `circ_pulse_def.schedule.duration` against `scheduled_circuit.instruction_durations.get(inst, qubits)` and raise an error if they differ.

Reply:
Excuse me for confusing you, I'm using `scheduled_circuit.instruction_durations` as like start_times of the scheduled circuit, i.e.  `scheduled_circuit.instruction_durations is not None` means the circuit is scheduled as well as `scheduled_circuit.duration is not None`. So setting `scheduled_circuit.instruction_durations` within `schedule()` is not good, but no problem for using `scheduled_circuit.instruction_durations` here.

Reply:
I suspect those issues come from the current design that sequence() calls schedule().
For example, `sequence` accepts a scheduled circuit while `schedule` accepts an ordinary circuit.
That implies scheduling is not a task of sequence. Scheduling should be done in transpiler (we are going to add scheduling passes for that purpose). (Although I temporarily leave the "measure rescheduling" logic in lower_gates to keep this PR simple, but I think such a functionality should be separated out as a transpiler pass in the future.)

Why don't we implement sequence() independent of schedule()?
I experimented it at f9f996b, and found that makes things clearer. What do you think?

Reply:
That commit seems fine, but it is not related to my question here, and it does not clarifies this point to me.

You're telling me that `duration` is not actually the duration of the instruction, but is the start time of the instruction? I don't understand

Reply:
Excuse me for your confusion, `instruction_durations` definitely keep durations, which is used in scheduling. I just meant durations and start times have the same amount of information so we could replace `instruction_durations` with `start_times` if we wanted. (I know we prefer durations)

To represent "scheduled circuit" by tweaking `QuantumCircuit` (instead of introducing a new subclass), I added two members `duration` and `instruction_durations` to it.
I first added only the `duration` but found the need of `instruction_durations` later.
Without `instruction_durations`, we cannot know start time for each instruction only by a scheduled circuit object. I think it is *necessary for simulator users* who want to define their own durations. I know it is redundant for real backends users. They can ask durations to a backend when reconstructing a schedule. Thinking about use cases for simulator users may help you understand my implementation.

```
# build a scheduled circuit without backend (for simulation)
sched_circ = transpile(qc, scheduling_method='alap', 
                       basis_gates=['u3', 'cx'], 
                       instruction_durations=[('u3', 0, 200), ('cx', [0, 1], 800)])
```

Reply:
Ah, we're talking about `sequence`, and the `sequence` is only used by real backend users. I'll think this again, whether we really need `QuantumCircuit.instruction_durations` or not.

Reply:
I've now understood your first suggestion. I've accepted your suggestion that we compute start times from`circ_pulse_defs` and removed `QuantumCircuit.instruction_durations` at cbe7ac2. It would be useful to add `instruction_durations` as an optional argument of `sequence` for checking the compatibility between `inst_map` and `instruction_durations`, but I'd like to leave it for future work.